File: ios/sampleNativeModules/SharedStorage.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(SharedStorage, NSObject)

RCT_EXTERN_METHOD(setItem:(NSString *)key
                  value:(NSString *)value
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getItem:(NSString *)key
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/sampleNativeModules/SharedStorage.swift
import Foundation
import WidgetKit

@objc(SharedStorage)
class SharedStorage: NSObject {

  let appGroup = "group.org.reactjs.native.example.yourapp"

  @objc(setItem:value:withResolver:withRejecter:)
  func setItem(key: String,
               value: String,
               resolve: @escaping RCTPromiseResolveBlock,
               reject: @escaping RCTPromiseRejectBlock) {

    guard let defaults = UserDefaults(suiteName: appGroup) else {
      reject("ERR", "App Group not found", nil)
      return
    }

    defaults.set(value, forKey: key)
    defaults.synchronize()

    WidgetCenter.shared.reloadAllTimelines()
    resolve(true)
  }

  @objc(getItem:withResolver:withRejecter:)
  func getItem(key: String,
               resolve: @escaping RCTPromiseResolveBlock,
               reject: @escaping RCTPromiseRejectBlock) {

    guard let defaults = UserDefaults(suiteName: appGroup) else {
      reject("ERR", "App Group not found", nil)
      return
    }

    let value = defaults.string(forKey: key)
    resolve(value)
  }

  @objc static func requiresMainQueueSetup() -> Bool {
    return false
  }
}